src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/ProjectExtensions.cs

```diff
@@ -58,6 +66,145 @@ public static async ValueTask<bool> AppliesToProjectAsync(this Type type, IProje
                 }
             }
 
+            // Check whether the type has an [DiagnosticAnalyzer] attribute
+            // If one exists, the type only applies to the project if the language matches
+            if (!AppliesToLanguage(type, project))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Checks to see if the type should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">A type that represents an upgrade-assistant feature.</param>
+        /// <param name="project">The project currently being upgraded.</param>
+        /// <returns>True if the object is described as matching the project's language. Objects not defning the language they support will be true as default.</returns>
+        private static bool AppliesToLanguage(Type type, IProject project)
+        {
+            // check to see if this is a code fixer
+            if (DoesCodeFixerFilterThisFeature(type, project))
+            {
+                return false;
+            }
+
+            // check to see if this is a configuration step or other feature using the ApplicableLanguageAttribute
+            if (DoesApplicableLanguageFilterThisFeature(type, project))
+            {
+                return false;
+            }
+
+            // if the attribute is not applied then this type defaults to opt into language support
+            return true;
+        }
+
+        /// <summary>
+        /// Checks to see if the type is a codefixer that should be filtered out of the user's view.
+        /// </summary>
+        /// <param name="type">A type that represents a codefixer <see cref="ExportCodeFixProviderAttribute"/>.</param>
+        /// <param name="project">The project currently being upgraded.</param>
+        /// <returns>True if the codefixer should be hidden from the user. Will return false for non-codefixer objects.</returns>
+        private static bool DoesCodeFixerFilterThisFeature(Type type, IProject project)
+        {
+            var analyzerAttr = type.CustomAttributes.FirstOrDefault(a => a.AttributeType.FullName.Equals(typeof(ExportCodeFixProviderAttribute).FullName, StringComparison.Ordinal));
+            if (analyzerAttr is null)
+            {
+                // this object is not the type we thought it was, the filter does not apply
+                return false;
+            }
+
+            var supportedLanguages = new List<string>();
+            var firstLanguage = analyzerAttr.ConstructorArguments.First().Value as string;
```

Comment:
This is probably unnecessary, but using FirstOrDefault might be safer here since I think it's preferable to have the attribute ignored than to throw an exception.